src/charm.py

```diff
@@ -691,19 +632,94 @@ def delete_file(self, path):
+        if not self.unit.is_leader():
+            return {}
+        if dbs := self.get_relation_databases():
+            return dbs
+
+        databases = {}
+        for relation in self.model.relations.get("db", []):
+            database = self.legacy_db_relation.get_databags(relation)[0].get("database")
+            if database:
+                databases[relation.id] = {
+                    "name": database,
+                    "legacy": True,
+                }
 
-    def render_pgb_config(self, config: PgbConfig, reload_pgbouncer=False) -> None:
+        for relation in self.model.relations.get("db-admin", []):
+            database = self.legacy_db_admin_relation.get_databags(relation)[0].get("database")
+            if database:
+                databases[relation.id] = {
+                    "name": database,
+                    "legacy": True,
+                }
+
+        for rel_id, data in self.client_relation.database_provides.fetch_relation_data(
+            fields=["database"]
+        ).items():
+            database = data.get("database")
+            if database:
+                databases[rel_id] = {
+                    "name": database,
+                    "legacy": False,
+                }
+        self.set_relation_databases(databases)
+        return databases
+
+    def _get_relation_config(self) -> [Dict[str, Dict[str, Union[str, bool]]]]:
+        """Generate pgb config for databases and admin users."""
+        if not self.backend.relation or not (databases := self.get_relation_databases()):
+            return {}
+
+        # In postgres, "endpoints" will only ever have one value. Other databases using the library
+        # can have more, but that's not planned for the postgres charm.
+        if not (postgres_endpoint := self.backend.postgres_databag.get("endpoints")):
+            return {}
+        host, port = postgres_endpoint.split(":")
+
+        read_only_endpoints = self.backend.get_read_only_endpoints()
+        r_hosts = ",".join([r_host.split(":")[0] for r_host in read_only_endpoints])
+        if r_hosts:
+            for r_host in read_only_endpoints:
+                r_port = r_host.split(":")[1]
+                break
+
+        pgb_dbs = {}
+
+        for rels_dbs in databases.values():
+            for database in rels_dbs:
+                name = database["name"]
+                pgb_dbs[name] = {
+                    "host": host,
+                    "dbname": name,
+                    "port": port,
+                    "auth_user": self.backend.auth_user,
+                }
+                if r_hosts:
+                    pgb_dbs[f"{name}_readonly"] = {
+                        "host": r_hosts,
+                        "dbname": name,
+                        "port": r_port,
+                        "auth_user": self.backend.auth_user,
+                    }
+                    if database["legacy"]:
+                        pgb_dbs[f"{name}_standby"] = pgb_dbs[f"{name}_readonly"]
```

Comment:
Is `_on_changed ` being removed in this case?

Reply:
Backend dbs will retrigger their own rerenders ATM. If we want to pass through Peers `_on_changed` we can raise flags from the backend db, but then we'll either have to put something random in the flag value or rerender twice (once to raise the flag and a second time to lower it).